crates/merkle-inclusion/src/lib.rs

```diff
@@ -0,0 +1,178 @@
+            // TODO better error
+            return Err(SynthesisError::Unsatisfiable);
+        }
+
+        Ok(Self {
+            leaf: Leaf::try_from(value[0..HASH_LENGTH * 2 - 1].to_owned())?,
+            siblings: value[HASH_LENGTH * 2..]
+                .chunks(HASH_LENGTH)
+                .map(|chunk| chunk.to_vec())
+                .collect(),
+        })
+    }
+}
+
+impl Proof {
+    pub fn leaf(&self) -> &Leaf {
+        &self.leaf
+    }
+
+    pub fn siblings(&self) -> &[Vec<Boolean>] {
+        &self.siblings
+    }
+}
+
+pub fn verify_proof<E, CS>(cs: CS, input: &[Boolean]) -> Result<Vec<Boolean>, SynthesisError>
+where
+    E: PrimeField,
+    CS: ConstraintSystem<E>,
+{
+    // Length of input is 256*3*nbr_siblings = 256 (expected_root) + 256 (proof_leaf_key) + 256 (proof_leaf_hash_value) + 256 * nbr_siblings
+    assert!(input.len() >= HASH_LENGTH * 4);
+    assert_eq!(input.len() % HASH_LENGTH, 0);
+
+    let expected_root = input[0..HASH_LENGTH - 1].to_vec();
+    let proof = Proof::try_from(input[HASH_LENGTH..].to_vec())?;
+
+    //Assert that we do not have more siblings than the length of our hash (otherwise cannot know which path to go)
+    // todo shouldn't this return proper error ?
+    assert!(
+        proof.siblings.len() <= HASH_LENGTH,
+        "Merkle Tree proof has more than {} ({}) siblings.",
+        HASH_LENGTH,
+        proof.siblings.len(),
+    );
+
+    // Reconstruct the root hash from the leaf and sibling hashes
+    let (cs, actual_root_hash) = construct_actual_hash(
+        cs,
+        proof.siblings().to_vec(),
+        proof.leaf().key().to_vec(),
+        proof.leaf().hash().to_vec(),
+    )?;
+
+    hash_equality(cs, expected_root, actual_root_hash)
+}
+
+fn construct_actual_hash<E, CS>(
+    mut cs: CS,
+    siblings: Vec<HashValue>,
+    leaf_key: HashValue,
+    leaf_hash: HashValue,
+) -> Result<(CS, HashValue), SynthesisError>
+where
+    E: PrimeField,
+    CS: ConstraintSystem<E>,
+{
+    let mut hash = leaf_hash;
+
+    for (i, (sibling_hash, bit)) in siblings
+        .iter()
+        .zip(leaf_key.iter().rev().skip(HASH_LENGTH - siblings.len()))
+        .enumerate()
+    {
+        let cs = &mut cs.namespace(|| format!("sibling {}", i));
+
+        if let Some(b) = bit.get_value() {
+            if b {
+                hash = hash_combine(cs, sibling_hash.to_owned(), hash)?
+            } else {
+                hash = hash_combine(cs, hash, sibling_hash.to_owned())?
+            }
+        } else {
+            // TODO better error
+            return Err(SynthesisError::Unsatisfiable);
+        }
+    }
+    Ok((cs, hash))
+}
+
+// Implementing hash_combine using SHA-256 to combine two hash values
+fn hash_combine<E, CS>(
+    mut cs: CS,
+    hash1: HashValue,
+    hash2: HashValue,
+) -> Result<HashValue, SynthesisError>
+where
+    E: PrimeField,
+    CS: ConstraintSystem<E>,
+{
+    sha3(cs.namespace(|| "hash combine"), &[hash1, hash2].concat())
```

Comment:
I may be missing context for this pull request. Can the `sha3` here be replaced with a generic hash function circuit? For example, someone may want to use Poseidon to construct a Merkle tree and prove inclusion.

Reply:
For sure, I'm currently working on a commit that's dedicated to it. Should be pushed quite soon :+1: